File: semester_4/numerical_methods/one_dimensional_optimization/dichotomy.py
from typing import Callable


def dichotomy_search(f: Callable, a: float, b: float, eps: float) -> float:
    while abs(a - b) > 2 * eps:
        x = (a + b) / 2
        x_1 = x - (eps / 2)
        x_2 = x + (eps / 2)
        if f(x_1) > f(x_2):
            a = x_1
        else:
            b = x_2

    return f((a + b) / 2)

File: semester_4/optimization/lw1_one_dimensional_optimization/src/dichotomy.rs
pub fn dichotomy_search<F>(f: F, a: f64, b: f64, eps: f64) -> (f64, f64)
where
    F: Fn(f64) -> f64,
{
    let mut a = a;
    let mut b = b;

    while (a - b).abs() > 2.0 * eps {
        let x = (a + b) / 2.0;
        let x_1 = x - (eps / 2.0);
        let x_2 = x + (eps / 2.0);

        if f(x_1) > f(x_2) {
            a = x_1;
        } else {
            b = x_2;
        }
    }

    ((a + b) / 2.0, f((a + b) / 2.0))
}
